Add option `providerOptional` to `createContext` to make warning optional

Download opened the pull request:
* Change noopEmitter to actually do nothing
* Add warnEmitter with same behavior as old noopEmitter
* Import the warnEmitter in addition to the noopEmitter
* Define an interface `Options` with one optional boolean field `providerOptional`
* Add an extra optional parameter `options` of type `Options` to `createContext`
* In `getEmitter` check the `providerOptional` flag and return `noopEmitter` i.s.o. `warnEmitter` for missing provider when it is set
* Add `warningOptional.spec.tsx`
* Test all variations of provider missing / available, optional / required

Fixes https://github.com/valotas/preact-context/issues/24

Download committed:
* Change noopEmitter to actually do nothing
* Add warnEmitter with same behavior as old noopEmitter
* Import the warnEmitter in addition to the noopEmitter
* Define an interface `Options` with one optional boolean field `providerOptional`
* Add an extra optional parameter `options` of type `Options` to `createContext`
* In `getEmitter` check the `providerOptional` flag and return `noopEmitter` i.s.o. `warnEmitter` for missing provider when it is set
* Add `warningOptional.spec.tsx`
* Test all variations of provider missing / available, optional / required

Download commented:
@valotas 

You already had an optional argument for `createContext` but I think you inherited this from the `React` api right? So I decided not to mess with this and add a third, extra optional argument `options`. This just has one option for now: `providerOptional: boolean`.

Let me know what you think of this or if I need to change something.

Download committed:
Fixed linter errors

valotas commented on src/context.ts:
```diff
@@ -133,7 +139,10 @@ function _createContext<T>(
-      return this.context[key] || noopEmitter;
+      return (
+        this.context[key] ||
+        ((options || {}).providerOptional ? noopEmitter : warnEmitter)
```
I believe it is better to create a `defaultEmitter` upon the context creation and here just `return this.context[key] || defaultEmitter;`

Download replied:
Hi @valotas I actually agree, but I did not see an easy way to do that. Because the emitter does not have access to the `providerOptional` flag. Or does it? Do you have a specific idea for how to implement this change?

Download replied:
Oh by the way, are you happy with the optional third parameter `options` and the `Options` interface? I must admit I am new to Typescript so I was not sure if that was the best way. 

I thought it best to opt (*sneaky pun*) for an options object instead of a single boolean param because maybe one day there will come more options... :)

valotas commented on src/_tests/warningOptional.Spec.tsx:
```diff
@@ -0,0 +1,174 @@
+import expect from "expect";
```
Why don't you just add the tests to the existing `context.Spec.tsx` in a different `describe` block?

Download replied:
Yeah I was not sure whether you wanted that. I saw that there were multiple test files already so I kinda assumed (I know, I know... never!) that you liked it this way.

Would you like me to change this?

valotas replied:
Well... You did it :+1:

valotas commented:
If you need any help regarding the changes, please let me know. I could also do them for you.

Download commented:
Hi @valotas If you could let me know whether to move the tests or not and maybe give some hint about how to consolidate both emitters into a `defaultEmitter` I will be happy to perform the changes you want.

:relaxed:

Download commented:
Mmmmm

> I believe it is better to create a defaultEmitter upon the context creation

I think I understand what you mean now... wrap the emittor creator in a function that accepts the `Options` object (or the boolean flag) and call that inside `createContext`, then use the created emitter.... I will take a stab at it right away.

Oh I have to explain about my personality here... I ask a lot of questions and then are too impatient to wait for the answers! Sorry!! lol!

Download committed:
Moved the tests from warningOptional.Spec.tsx to context.Spec.tsx. One test proved to be redundant; it was already tested in context.Spec.tsx so I just removed it. The tests in context.Spec.tsx were good examples and following them allowed me to greatly simplify the test code.

Download committed:
Grrr forgot to run linter, fixed some errors :)

Download committed:
Merge noopEmitter and warnEmitter into defaultEmitter.

Removed `noopEmitter` and `warnEmitter` from context-value-emitter.ts and introduced `defaultEmitter` in context.ts which is created inside the `createContext` function and uses the `providerOptional` flag to determine whether to log a warning if there is no provider.

Download commented:
Hi @valotas 

I think I implemented the changes you requested. Thanks for your feedback, I think the code improved because of it.

If there is more you want changed, please let me know.

-Stijn

valotas commented on src/context-value-emitter.ts:
```diff
@@ -38,15 +38,3 @@ export function createEmitter<T>(
   };
 }
-
-export const noopEmitter: ContextValueEmitter<any> = {
```
This file is the natural place for any emitter. You should not move this.

valotas replied:
I believe you need a factory here instead of the noopEmitter instance

Download replied:
> This file is the natural place for any emitter. You should not move this.

Agreed. But it does not have access to the boolean flag to check

> I believe you need a factory here instead of the noopEmitter instance

I guess you are saying I should wrap the class in a factory function that accepts the boolean flag?

This was my initial idea but I did not like the overhead of the extra function... Plus.. passing the boolean directly is ugly. It would be better if the emitter factory function accepts the `Options` interface... which is in the other file (and does look like it belongs there) which means we get a circular dependency...

So maybe you can elaborate a bit on how you would like to see this and then I can make the change as you wish.

valotas commented:
I'll have a closer look at it this weekend. Thanks again!

valotas commented:
Now that I think of it, wouldn't it be better to just have a warn function as an option? That way someone can choose how to handle warnings. Even through an exception if you feel like being strict about it.

Download commented:
> wouldn't it be better to just have a warn function as an option?

Yeah.... I was actually thinking of creating a second PR after this one that would introduce a `logger` option so you can configure the logging handling  👍 

For me, whether `preact-context` should warn or not is one thing and how to handle logging messages is another. Take my use case for example. I want to inform 'preact-context' that in my scenario, not having a provider is a-ok and expected and not a reason to warn. When a warning does come, I would like to have it logged via the logging library I am using throughout the app. Without the `providerOptional` flag I would have to do this through implementing the `warn()`  function, I would have to resort to parsing the message and in case it is `'Consumer used without a Provider'` then suppress it, otherwise log it as usual. This is ugly imho.

So my idea here would be (for my scenario):

```js
const context = createContext({}, undefined, {
  providerOptional: true,
  logger: myLoggerLib
})
```

What do you think?

Download committed:
Making warning optional and logger configurable. Move emitter back.

+ Added interface `Log` with method `warn`
+ Added field `log` to interface `Options`
+ Moved interfaces `Log` and `Options` to context-value-emitter.ts
+ Moved `defaultEmitter` back to context-value-emitter.ts
+ Wrapped `defaultEmitter` in a factory function that accepts `Options`
+ Call `createDefaultEmitter(options)` in context.ts
+ Made context.ts export `Options`
+ Added tests for overridden logger and arguments to `createContext`

Download commented on package.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "preact-context",
+  "name": "@stijndewitt/preact-context",
```
Oops

Download committed:
oops, I had published an intermediate version to my private scope, but that should not be in here

Download committed:
import Options as well as export it

Download commented:
I think maybe this is the one to use... But there are lots of commits so please don't merge this PR... I will create a new PR with just a single commit.
